File: ios/FTZTextInputWithLoggerView.m
#import "FTZTextInputWithLoggerView.h"
#import "LoggingKeyboardView.h"

@implementation FTZTextInputWithLoggerView
{
  UITextField* textField;
  LoggingKeyboardView* customizedInputView;
}
@synthesize value;

- (instancetype)initWithBridge:(RCTBridge *)bridge
{
  if(self = [super initWithBridge:bridge]) {
    // initialize custom keyboard input view
    customizedInputView= [[LoggingKeyboardView alloc] init];

    // grab textview from textinput view and attach custom keyboard input view to it
    textField = (UITextField*)self.backedTextInputView;
    textField.delegate = self;
    [textField setInputView:customizedInputView];
    [(UITextField*)textField addTarget:self
      action:@selector(textFieldDidChange)
      forControlEvents:UIControlEventEditingChanged
    ];

    // add reference to textview from inputview so input knows where to send its actions
    [customizedInputView setTextField:textField];
    customizedInputView.delegate = self;
  }
  return self;
}

- (void)setAutocompleteLabel:(NSString *)incomingValue
{
  NSLog(@"C00: set auto complete");
  customizedInputView.autocompleteLabel.text = incomingValue;
}

- (void)setAutocompleteValue:(NSNumber *)incomingValue
{
  NSLog(@"G00: set autcomplete value");
  [customizedInputView.autocompleteButton setTitle:[incomingValue stringValue] forState:UIControlStateNormal];
  [customizedInputView.autocompleteButton setEnabled:YES];
  customizedInputView.autocompleteValue = incomingValue;
}

- (void)setUnit:(NSString *)unit
{
  customizedInputView.unitLabel.text = unit;
}

- (void)setStepValue:(NSNumber *)value
{
  [customizedInputView.incrementButton setTitle:[NSString stringWithFormat:@"+%@",[value stringValue]] forState:UIControlStateNormal];
  [customizedInputView.decrementButton setTitle:[NSString stringWithFormat:@"-%@",[value stringValue]] forState:UIControlStateNormal];
  customizedInputView.stepValue = value;
}

- (void)setValue:(NSString *)incomingValue
{
  NSLog(@"A00: Received value %@", incomingValue);
  NSLog(@"A00: textField value: %@", textField.text);
  if([incomingValue isEqualToString:textField.text]){
    return;
  }

  NSLog(@"A00: overriding");
  textField.text = incomingValue;
  value=incomingValue;
  [textField sendActionsForControlEvents:UIControlEventEditingChanged];
}

- (void)textFieldDidChange
{
  NSLog(@"A00: text field changed!!! %@", textField.text);

  if([textField.text isEqualToString:self.value] || !self.onChangeText){
    return;
  }
  NSLog(@"A00: Output value %@", textField.text);
  self.onChangeText(@{
    @"text": textField.text
  });
}

-(void) setLeftButtonEnabled:(BOOL)enabled
{
  if(enabled){
    [customizedInputView.leftButton setEnabled:YES];
  } else {
    [customizedInputView.leftButton setEnabled:NO];
  }
}

-(void) setRightButtonEnabled:(BOOL)enabled
{
  if(enabled){
    [customizedInputView.rightButton setEnabled:YES];
  } else {
    [customizedInputView.rightButton setEnabled:NO];
  }
}

#pragma mark LoggingKeyboardViewDelegate
- (void) leftButtonPressed
{
  if(!self.onLeftButtonPress){
    return;
  }
  self.onLeftButtonPress(@{});
}

- (void) rightButtonPressed
{
  if(!self.onRightButtonPress){
    return;
  }
  self.onRightButtonPress(@{});
}

#pragma mark UITextFieldDelegate
- (BOOL) textFieldShouldBeginEditing:(UITextField *)textField {
  dispatch_async(dispatch_get_main_queue(), ^{
    [textField selectAll:nil];
  });
  return YES;
}

- (BOOL) textFieldDidBeginEditing:(UITextField *)textField {
  if(self.onFocus){
    self.onFocus(@{});
  }
  return YES;
}

- (BOOL) textFieldDidEndEditing:(UITextField *)textField {
  if(self.onBlur){
    self.onBlur(@{});
  }
  return YES;
}

@end

File: ios/FTZTextInputWithLoggerView.swift
class FTZTextInputWithLoggerView: RCTSinglelineTextInputView,
UITextFieldDelegate, FTZCustomInputViewDelegate {

    weak var textField: UITextField?
    var customInputView: FTZCustomInputView?

    override init(bridge: RCTBridge) {
        super.init(bridge: bridge)
        let textInputView = self.backedTextInputView as! UITextField

        customInputView = FTZCustomInputView(targetTextField: textInputView);
        customInputView!.delegate = self
        textInputView.inputView = customInputView
    }

    override func didMoveToSuperview() {
        textField = self.backedTextInputView as? UITextField
        textField?.delegate = self;
    }

    @objc func setIsLeftButtonDisabled(_ val: ObjCBool) {
        customInputView?.leftButton?.isEnabled = !val.boolValue
    }

    @objc func setIsRightButtonDisabled(_ val: ObjCBool) {
        customInputView?.rightButton?.isEnabled = !val.boolValue
    }

    @objc func setStepValue(_ val: NSNumber) {
        let stepValue = val.doubleValue
        let formatter = NumberFormatter()
        formatter.minimumFractionDigits = 0
        formatter.maximumFractionDigits = 3
        let label = formatter.string(from: val)
        if (label == nil) {
            return
        }
        customInputView?.incrementButton?.value = stepValue;
        customInputView?.incrementButton?.setTitle("+" + label!, for: .normal)
        customInputView?.decrementButton?.value = -1 * stepValue;
        customInputView?.decrementButton?.setTitle("-" + label!, for: .normal)
    }

    @objc func setSuggestLabel(_ val: NSString) {
        customInputView?.suggestLabel?.text = val as String
    }

    @objc func setSuggestValue(_ val: NSNumber) {
        let formatter = NumberFormatter()
        formatter.minimumFractionDigits = 0
        formatter.maximumFractionDigits = 3
        let label = formatter.string(from: val)
        customInputView?.suggestButton?.isEnabled = true
        customInputView?.unitLabel?.isEnabled = true
        customInputView?.suggestButton?.setTitle(label, for: .normal)
    }

    @objc func setUnitLabel(_ val: NSString) {
        customInputView?.unitLabel?.text = val as String
    }

    // MARK: - UITextFieldDelegate

    @objc var onFocus: RCTDirectEventBlock?
    func textFieldDidBeginEditing(_ textField: UITextField) {
        if onFocus != nil {
            onFocus!(nil)
        }
        DispatchQueue.main.async {
            textField.selectAll(nil);
        }
    }

    @objc var onBlur: RCTDirectEventBlock?
    func textFieldDidEndEditing(_ textField: UITextField) {
        if onBlur != nil {
            onBlur!(nil)
        }
    }

    // MARK: - FTZCustomInputViewDelegate

    @objc var onLeftButtonPress: RCTDirectEventBlock?
    func leftButtonPressed() {
        if onLeftButtonPress != nil {
            onLeftButtonPress!(nil)
        }
    }

    @objc var onRightButtonPress: RCTDirectEventBlock?
    func rightButtonPressed() {
        if onRightButtonPress != nil {
            onRightButtonPress!(nil)
        }
    }
}

@objc (FTZTextInputWithLoggerManager)
class FTZTextInputWithLoggerManager: RCTSinglelineTextInputViewManager {
 
  override static func requiresMainQueueSetup() -> Bool {
    return true
  }
 
  override func view() -> UIView! {
    return FTZTextInputWithLoggerView(bridge: self.bridge)
  }
 
}
